assets/js/SignDisplay.tsx

```diff
@@ -0,0 +1,60 @@
+import * as React from 'react';
+import SignText from './SignText';
+import { SingleSignContent } from './types';
+import { lineDisplayText } from './helpers';
+
+interface SignDisplayProps {
+  content: SingleSignContent;
+  currentTime: number;
+}
+
+type Event = {
+  timestamp: number;
+  value: { top: string; bottom: string; duration: number } | null;
+};
+
+function SignDisplay({ currentTime, content }: SignDisplayProps) {
+  const [initialTime] = React.useState(currentTime);
+
+  // Compute a time-sequence of page transitions from recent audio-visual
+  // messages, and find which one we should be on.
+  const currentPage = content.audios
+    .filter((audio) => audio.visual_data)
+    .reduce<Event[]>(
+      (events, audio) => {
+        const startTimestamp = Math.max(
+          events.at(-1)?.timestamp ?? 0,
+          audio.timestamp,
+        );
+        const [newEvents, newTimestamp] = audio.visual_data.pages.reduce<
+          [Event[], number]
+        >(
+          ([innerEvents, timestamp], page) => [
+            [...innerEvents, { timestamp, value: page }],
+            timestamp + page.duration * 1000,
+          ],
+          [events, startTimestamp],
+        );
+        return [...newEvents, { timestamp: newTimestamp, value: null }];
+      },
+      [{ timestamp: 0, value: null }],
+    )
```

Comment:
I found this reduce a little hard to follow on without taking a few passes at it. Maybe we could break out the inner reduce out to a helper to help readability? I also found the term `Event` to be a bit ambiguous in this context and I think something slightly more descriptive could be helpful. Even just prepending `Visual` to make it `VisualEvent` or something?

Reply:
I split out the inner function and decoupled the logic a little. Let me know what you think.